bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSource.java

```diff
@@ -0,0 +1,186 @@
+                    }
+                    setIdle(true);
+                    // get raw audio from the pulse audio socket
+                    return new PulseAudioStream(sourceFormat, this::getSourceInputStream, (idle) -> {
+                        setIdle(idle);
+                        if (idle) {
+                            scheduleDisconnect();
+                        }
+                    });
+                } catch (IOException e) {
+                    disconnect(); // disconnect force to clear connection in case of socket not cleanly shutdown
+                    if (countAttempt == 2) { // we won't retry : log and quit
+                        if (logger.isWarnEnabled()) {
+                            String port = clientSocket != null ? Integer.toString(clientSocket.getPort()) : "unknown";
+                            logger.warn(
+                                    "Error while trying to get audio from pulseaudio audio source. Cannot connect to {}:{}, error: {}",
+                                    pulseaudioHandler.getHost(), port, e.getMessage());
+                        }
+                        setIdle(true);
+                        throw e;
+                    }
+                } catch (InterruptedException ie) {
+                    logger.info("Interrupted during source audio connection: {}", ie.getMessage());
+                    setIdle(true);
+                    throw new AudioException(ie);
+                }
+                countAttempt++;
+            }
+        } catch (IOException e) {
+            throw new AudioException(e);
+        } finally {
+            scheduleDisconnect();
+        }
+        setIdle(true);
+        throw new AudioException("Unable to create input stream");
+    }
+
+    private @Nullable InputStream getSourceInputStream() {
+        try {
+            connectIfNeeded();
+        } catch (IOException | InterruptedException ignored) {
+        }
+        try {
+            return (clientSocket != null) ? clientSocket.getInputStream() : null;
+        } catch (IOException ignored) {
+            return null;
+        }
+    }
+
+    static class PulseAudioStream extends AudioStream {
+        private final Logger logger = LoggerFactory.getLogger(PulseAudioAudioSource.class);
+        private final AudioFormat format;
+        private final Supplier<@Nullable InputStream> getInput;
+        private final Consumer<Boolean> setIdle;
+
+        public PulseAudioStream(AudioFormat format, Supplier<@Nullable InputStream> getInput,
+                Consumer<Boolean> setIdle) {
+            this.getInput = getInput;
+            this.format = format;
+            this.setIdle = setIdle;
+        }
+
+        @Override
+        public AudioFormat getFormat() {
+            return format;
+        }
+
+        @Override
+        public int read() throws IOException {
+            logger.trace("reading from pulseaudio stream");
```

Comment:
You have 3 times the same log message in 3 different methods.
I would suggest to keep only one, the third one as it will be always be called finally.
Or if you want to have a log in each method, make a difference in the log message.

Reply:
I didn't realize that was happening, there is also an extra call to setIdle because of the same, thanks!